eval/src/vespa/eval/tensor/dense/dense_tensor_address_combiner.h

```diff
@@ -71,4 +94,81 @@ class DenseTensorAddressCombiner
     static eval::ValueType combineDimensions(const eval::ValueType &lhs, const eval::ValueType &rhs);
 };
 
+
+/**
+ * Utility class to iterate over common cells in a dense tensor.
+ */
+class CommonDenseTensorCellsIterator
+{
+public:
+    using size_type = eval::ValueType::Dimension::size_type;
+    using Address = std::vector<size_type>;
+    using Mapping = DenseTensorAddressCombiner::Mapping;
+private:
+    using Dims = std::vector<uint32_t>;
+    using CellsRef = vespalib::ConstArrayRef<double>;
+    const eval::ValueType &_type;
+    CellsRef               _cells;
+    Address                _address;
+    const Mapping         &_common;
+    const Mapping         &_mutable;
+    std::vector<size_t>    _accumulatedSize;
+
+    double cell(size_t cellIdx) const { return _cells[cellIdx]; }
+    size_t index(const Address &address) const {
+        size_t cellIdx(0);
+        for (uint32_t i(0); i < address.size(); i++) {
+            cellIdx += address[i]*_accumulatedSize[i];
+        }
+        return cellIdx;
+    }
+public:
+    CommonDenseTensorCellsIterator(const Mapping & common, const Mapping & right,
+                                   const eval::ValueType &type_in, CellsRef cells);
+    ~CommonDenseTensorCellsIterator();
+    template <typename Func>
+    void for_each(Address & combined, Func && func) {
```

Comment:
writing directly into the address of the combiner totally undermines the job of the combiner... Also, with 'combined' address being passed around, it gets hard to keep track of which ones are complete and which are incomplete.